cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/cache.go

```diff
@@ -0,0 +1,324 @@
+/*
+Copyright © 2021 VMware
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package main
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/go-redis/redis/v8"
+	corev1 "github.com/kubeapps/kubeapps/cmd/kubeapps-apis/gen/core/packages/v1alpha1"
+	"github.com/kubeapps/kubeapps/cmd/kubeapps-apis/server"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	log "k8s.io/klog/v2"
+)
+
+type fluxPlugInCache struct {
+	// to prevent multiple watchers
+	watcherStarted bool
+	// this mutex guards watcherStarted var
+	watcherMutex sync.Mutex
+	redisCli     *redis.Client
+	clientGetter server.KubernetesClientGetter
+	// these waitGroup is used exclusively by unit tests to block until all expected repos have
+	// been indexed by the go routine running in the background. The creation of the WaitGroup object
+	// and .Add() is expected to be done by the unit test client. The server-side only signals
+	// when indexing one repo is complete
+	indexRepoWaitGroup *sync.WaitGroup
+}
+
+func newCache(clientGetter server.KubernetesClientGetter) (*fluxPlugInCache, error) {
+	log.Infof("+newCache")
+	REDIS_ADDR, ok := os.LookupEnv("REDIS_ADDR")
```

Comment:
I will add a comment so I don't forget

Reply:
I did add a comment in the code, as promised. Just wanted to also mention that on a second look, I don't necessarily agree. The way it is right now, is one place that reads env vars. What you're suggesting is to move this code out to server.go (kind of like helm plug-in server.go reads ASSET_SYNCER_* env vars today) or main.go? So then we'd each plug-in that wants cache read these env vars? That doesn't seem right. I think I am missing something. Anyway, we'll discuss this on a call I am sure.

Reply:
`cmd/root.go` - as in, the root CLI command would read all config on startup. If you look there, you'll see we already do this for the clusters config path and pinniped proxy URL, populating a server options which is passed on to `server.go` where the server is started. Note that the library we're using there (cobra/viper) [is already reading in env vars that match the configured options](https://github.com/kubeapps/kubeapps/blob/master/cmd/kubeapps-apis/cmd/root.go#L83), so you can supply a `--pinniped-proxy-url` or you can specify it in the env, either way, it'll end up in the value for the option.

> So then we'd each plug-in that wants cache read these env vars? That doesn't seem right.

No, each plugin would just be passed a function or interface that is already configured with the env vars. Not suggesting this can be done now, just when it's generalised for all plugins to use. This is much the same way that we currently [create a function closure to be passed to plugins](https://github.com/kubeapps/kubeapps/blob/master/cmd/kubeapps-apis/server/plugins.go#L256) so they can create a k8s client without needing to know anything about the pinniped proxy configuration.

> I think I am missing something. Anyway, we'll discuss this on a call I am sure.

Yep, and it won't be relevant until we're generalising the caching work for all plugins anyway.

Reply:
ok, sounds good